File: tests/fixtures/kat/erlang/taxonomy.erl
%% Comprehensive Erlang feature taxonomy fixture.

-module(taxonomy).

-export([add/2, branchy/2, describe/1, bark/2, use_animal/1]).

-define(CONSTANT, 42).

-record(animal, {name, species = "unknown"}).
-record(dog, {base, breed}).

add(A, B) ->
    A + B.

branchy(Value, Mode) when Value > 0 ->
    Total = Value + sum_even(Value),
    Reduced = reduce(Total),
    parse_or_error(Mode, Reduced);
branchy(Value, Mode) when Value < 0 ->
    Total = -Value + sum_even(Value),
    Reduced = reduce(Total),
    parse_or_error(Mode, Reduced);
branchy(_Value, Mode) ->
    parse_or_error(Mode, 0).

sum_even(Value) ->
    lists:sum([I || I <- lists:seq(0, Value - 1), I rem 2 =:= 0]).

reduce(N) when N > 100 ->
    reduce(N div 2);
reduce(N) ->
    N.

parse_or_error(Mode, Total) ->
    case string:to_integer(Mode) of
        {Int, _} when is_integer(Int) -> {ok, Int + Total};
        _ -> {error, "bad mode"}
    end.

new_animal(Name) ->
    #animal{name = Name}.

describe(#animal{name = Name, species = Species}) ->
    Name ++ " the " ++ Species.

new_dog(Name, Breed) ->
    Base = (new_animal(Name))#animal{species = "dog"},
    #dog{base = Base, breed = Breed}.

bark(#dog{base = #animal{name = Name}}, Loud) ->
    case Loud of
        true -> {error, "too loud"};
        false -> {ok, Name ++ ": woof"}
    end.

use_animal(Dog) ->
    bark(Dog, false).

File: tests/fixtures/kat/go/taxonomy.go
// Comprehensive Go feature taxonomy fixture.
// Exercises funcs, methods, structs, interfaces, branches, panics, and calls.

package taxonomy

import (
	"errors"
	"fmt"
	"strconv"
)

const Constant = 42

type Greeter interface {
	Greet(name string) string
}

type Status int

const (
	StatusPending Status = iota
	StatusDone
)

func Add(a, b int) int {
	return a + b
}

func Branchy(value int, mode string) (int, error) {
	total := 0
	if value > 0 {
		total += value
	} else if value < 0 {
		total -= value
	}
	for i := 0; i < value; i++ {
		if i%2 == 0 {
			total += i
		}
	}
	for total > 100 {
		total /= 2
	}
	parsed, err := strconv.Atoi(mode)
	if err != nil {
		return 0, errors.New("bad mode")
	}
	return parsed + total, nil
}

type Animal struct {
	Name    string
	Species string
}

func NewAnimal(name string) *Animal {
	return &Animal{Name: name, Species: "unknown"}
}

func (a *Animal) Describe() string {
	return fmt.Sprintf("%s the %s", a.Name, a.Species)
}

type Dog struct {
	*Animal
	Breed string
}

func NewDog(name, breed string) *Dog {
	base := NewAnimal(name)
	base.Species = "dog"
	return &Dog{Animal: base, Breed: breed}
}

func (d *Dog) Bark(loud bool) (string, error) {
	if loud {
		return "", errors.New("too loud")
	}
	return fmt.Sprintf("%s: woof", d.Name), nil
}

func (d *Dog) Greet(name string) string {
	return fmt.Sprintf("%s greets %s", d.Name, name)
}

func UseAnimal(d *Dog) (string, error) {
	return d.Bark(false)
}
